### PR TITLE
Small nano malloc cleanups

### PR DESCRIPTION
**Pull request by arichardson**
As an added side-effect the code is now compatible with CHERI-enabled architecture (Arm Morello, CHERI-RISC-V), although full support for these architectures requires changes to more files.

### CODE REVIEW

File: newlib/libc/stdlib/nano-mallocr.c [hunk 1]
```diff
@@ -55,33 +56,15 @@
...
+    void *p;
+    double d;
+    long long ll;
+    size_t s;
 } align_chunk_t;
 
 typedef struct {
-    char c;
-    size_t s;
+    uintptr_t s;
```
Comment (keith-packard) [inline newlib/libc/stdlib/nano-mallocr.c:67]: Why the change in type here? I have this vague memory of platforms where size_t is smaller than uintptr_t.

File: newlib/libc/stdlib/nano-mallocr.c [hunk 2]
```diff
@@ -102,27 +85,27 @@ typedef struct malloc_chunk
...
+    uintptr_t size;
 
     /* pointer to next chunk */
     struct malloc_chunk * next;
 } chunk_t;
 
 /* Alignment of allocated chunk. Compute the alignment required from a
  * range of types */
-#define MALLOC_CHUNK_ALIGN	(offsetof(align_chunk_t, u))
+#define MALLOC_CHUNK_ALIGN	_Alignof(align_chunk_t)
```
Comment (keith-packard) [inline newlib/libc/stdlib/nano-mallocr.c:96]: Thanks, switching to _Alignof is a nice cleanup.

File: newlib/libc/stdlib/nano-mallocr.c [hunk 3]
```diff
@@ -102,27 +85,27 @@ typedef struct malloc_chunk
...
 
 /* Alignment of the header. Never larger than MALLOC_CHUNK_ALIGN */
-#define MALLOC_HEAD_ALIGN	(offsetof(align_head_t, s))
+#define MALLOC_HEAD_ALIGN	_Alignof(align_head_t)
 
 /* Size of malloc header. Keep it aligned. */
-#define MALLOC_HEAD 		ALIGN_TO(sizeof(size_t), MALLOC_HEAD_ALIGN)
+#define MALLOC_HEAD 		__align_up(sizeof(uintptr_t), MALLOC_HEAD_ALIGN)
```
Comment (keith-packard) [inline newlib/libc/stdlib/nano-mallocr.c:102]: Thanks for using the cdefs.h macro here.

**Review by keith-packard — CHANGES_REQUESTED** [review]
Nice cleanup; can you switch the size field back to size_t? Or is there some reason to use uintptr_t for that?

File: newlib/libc/stdlib/nano-mallocr.c (hunk 1 above)
Comment (arichardson) [reply newlib/libc/stdlib/nano-mallocr.c:67]: That's exactly the type of platform I made this change for, for CHERI-enabled platforms size_t is half the size of pointers, but malloc still needs to return pointer-aligned values.

I believe it should be possible to use size_t instead of uintptr_t here, but having the header be less aligned than the malloc_chunk requires some additional non-trivial changes to the remaining code. I went for the simplest approach here which is to use uintptr_t which should be no-op change for all other platforms.
It's been a while since I made these changes, but here is what I wrote in the commit message back then:
> For CHERI size_t is half the size of uintptr_t (and thus pointers), so
using uinptr_t instead of size_t ensures there is no padding between
the malloc_chunk members `size` and `next`. This is required by the
remaining logic since otherwise the `next` pointer could become
unaligned if we only subtract MALLOC_HEAD (size_t) instead of a pointer
size quantity. Further changes would allow using a single size_t before
the actual allocation, but this minimally invasive change has no effect
on existing architectures while enabling the malloc test to pass for
CHERI purecap architectures.

Comment (keith-packard) [reply newlib/libc/stdlib/nano-mallocr.c:67]: Oh, it's actually designed to let the header be allocated on a less strict alignment than the data. For instance, on a regular 32-bit platform, size_t is 32 bits but malloc wants to return blocks aligned on 64-byte boundaries for long long and double values. If you allocate 12 bytes, there's a nice 4-byte chunk after your block but before the next 64-bit boundary that the size can fit into.

Of course, on that platform, uintptr_t and size_t are both 32-bits, so there's no impact. With Cheri, we should be able to get the same result by using the smaller size_t type?

**Comment by rdiez** [comment]
I looked into this allocator, but it's hard to understand, so I have a question. The old structure looks like this:

```
typedef struct {
    char c;
    size_t s;
} align_head_t;
```
The modified one is:

```
typedef struct {
    uintptr_t s;
} align_head_t;

```
Struct align_chunk_t lost "char c" too.

The question is: what was "char c" for?

About the modified "uintptr_t s" field: most people would expect a size_t there. If uintptr_t is preferable under some circumstances, I would state the reason next to the field. More often than not, you look at the source code and have no easy access to the original commit history. In fact, the history may get lost if code is copied from a different project.

In the rare scenario that uintptr_t is greater than size_t, perhaps for some quirky 8-bit architecture, is this change not going to waste memory?

**Comment by keith-packard** [comment]
> I looked into this allocator, but it's hard to understand, so I have a question. The old structure looks like this:
> 
> ```
> typedef struct {
>     char c;
>     size_t s;
> } align_head_t;
> ```
> 
> The modified one is:
> 
> ```
> typedef struct {
>     uintptr_t s;
> } align_head_t;
> ```
> 
> Struct align_chunk_t lost "char c" too.
> 
> The question is: what was "char c" for?

This struct was declared solely to discover the alignment required for 's' -- by creating a struct starting with 'char', the next field will be offset by the alignment required by that member. Using _Alignof is way clearer here.

> In the rare scenario that uintptr_t is greater than size_t, perhaps for some quirky 8-bit architecture, is this change not going to waste memory?

That's also my thinking. The other 'clever hack' within this code is that it allows the 's' field to be less strictly aligned than the overall allocation, which starts just past that.

File: newlib/libc/stdlib/nano-mallocr.c (hunk 1 above)
Comment (arichardson) [reply newlib/libc/stdlib/nano-mallocr.c:67]: That's a good point, it should only really matter for platforms where malloc alignment is > pointer size, and in that case size_t will be the same as uintptr_t, so I think using uintptr_t is still a no-op change?

**Comment by rdiez** [comment]
> This struct was declared solely to discover the alignment required for 's' --
> by creating a struct starting with 'char', the next field will be offset
> by the alignment required by that member.

I see. Thanks for the explanation.

> Using _Alignof is way clearer here.
> [...]
> That's also my thinking. 

I would then separate such clean-ups from the other change (size_t -> uintptr_t) and apply the clean-ups upfront.

File: newlib/libc/stdlib/nano-mallocr.c (hunk 1 above)
Comment (rdiez) [reply newlib/libc/stdlib/nano-mallocr.c:67]: > That's a good point, it should only really matter for platforms where malloc alignment is
> pointer size, and in that case size_t will be the same as uintptr_t,
> so I think using uintptr_t is still a no-op change?

Most people would expect a size_t there. Could you please explain again in which scenario the type uintptr_t is better than size_t?

Comment (keith-packard) [reply newlib/libc/stdlib/nano-mallocr.c:67]: sure, `uintptr_t` and `size_t` are generally the same underlying integer type. However, using `uintptr_t` instead of `size_t` is misleading about what this field holds -- it doesn't hold an integer representation of a pointer value, it holds the size of the allocation, and that is defined in the interface to `malloc` as `size_t`.

Is there any reason that you need to use `uintptr_t` instead of `size_t` for this field?

**Comment by keith-packard** [comment]
> I would then separate such clean-ups from the other change (size_t -> uintptr_t) and apply the clean-ups upfront.

I still don't see any motivation for that particular change; maybe we're missing something though?

**Comment by rdiez** [comment]
> I still don't see any motivation for that particular change; maybe we're missing something though?

But the clean-ups alone are an improvement by themselves which have nothing to do with the type change, or am I missing something?

**Comment by keith-packard** [comment]
> I still don't see any motivation for that particular change; maybe we're missing something though?

Oh, looks like I didn't read all of the responses carefully enough. On Cheri, pointers are larger than `size_t`, and get a larger alignment value as well. That means `struct malloc_chunk` gets padding inserted between `size_t size` and `struct malloc_chunk *next`. That's going to cause problems because the code assumes the `size` field is immediately followed by the `next` field, without padding. I'll ponder what the best solution for this might be which preserves the correct type for `size` while still ensuring that `size` is immediately followed by the application visible memory space.

**Comment by keith-packard** [comment]
ok, I think a reasonable solution here is to use `uintptr_t` and insert a comment about how that ensures that the `size` field and `chunk` field are the same size and require the same alignment, and that while `size_t` might be more correct from a language perspective, that causes problems when `size_t` is smaller than `struct malloc_chunk *next`.

**Comment by keith-packard** [comment]
> But the clean-ups alone are an improvement by themselves which have nothing to do with the type change, or am I missing something?

You're right -- the cleanups are quite nice and should get merged. But, I think @arichardson was motivated to make these changes as a part of getting the code running on Cheri, which requires the type change. Now that I think I understand that type change, I think it's actually an additional 'clean up' as the existing code makes inappropriate assumptions about data type sizes and alignments.

Hrm. The proposed change (`size_t` -> `uintptr_t`) assumes that `uintptr_t` and `struct malloc_chunk *` have the same size. That's likely true in any "reasonable" C implementation, but I don't find any assurances about that in the standard. I wonder if a better solution wouldn't be to use a union of `size_t` with `struct malloc_chunk *`. Something like this:
```
diff --git a/newlib/libc/stdlib/nano-mallocr.c b/newlib/libc/stdlib/nano-mallocr.c
index a537d42f7b..5ce23027de 100644
--- a/newlib/libc/stdlib/nano-mallocr.c
+++ b/newlib/libc/stdlib/nano-mallocr.c
@@ -79,10 +79,8 @@ typedef struct {
     } u;
 } align_chunk_t;
 
-typedef struct {
-    char c;
-    size_t s;
-} align_head_t;
+#define ALIGN_HEAD_T union { size_t size; struct malloc_chunk *pointer; }
+typedef ALIGN_HEAD_T align_head_t;
 
 typedef struct malloc_chunk
 {
@@ -101,8 +99,8 @@ typedef struct malloc_chunk
      * aligned correctly
      */
 
-    /* size of the allocated payload area */
-    size_t size;
+    /* size of the allocated payload area. */
+    ALIGN_HEAD_T;
 
     /* pointer to next chunk */
     struct malloc_chunk * next;
@@ -113,7 +111,7 @@ typedef struct malloc_chunk
 #define MALLOC_CHUNK_ALIGN     (offsetof(align_chunk_t, u))
 
 /* Alignment of the header. Never larger than MALLOC_CHUNK_ALIGN */
-#define MALLOC_HEAD_ALIGN      (offsetof(align_head_t, s))
+#define MALLOC_HEAD_ALIGN      (_Alignof(align_head_t))
 
 /* Size of malloc header. Keep it aligned. */
 #define MALLOC_HEAD            ALIGN_TO(sizeof(size_t), MALLOC_HEAD_ALIGN)
```
This uses `#define ALIGN_HEAD_T` so that we end up with an anonymous union allowing direct access to the `size` field, minimizing changes in the rest of the file.

**Comment by rdiez** [comment]
> That means struct malloc_chunk gets padding inserted
> between size_t size and struct malloc_chunk *next.
> [...]
> ok, I think a reasonable solution here is to use `uintptr_t`

I still don't quite grasp it. Inserting padding would potentially waste memory, wouldn't it? Let's say the size and alignment for a pointer is 64 bits, and the size and alignment for size_t are 32 bits. Shouldn't the aim be to avoid the automatic padding? If the 64-bit pointer is aligned, the 32-bit size_t immediately before it would be sufficiently aligned too, right?

I'm thinking of some weird 8-bit or 16-bit architecture with much more ROM/Flash than RAM. Pointers may be longer, in order to accommodate for the different memory areas, but dynamic allocations in RAM could be much smaller. I have STM32 microcontrollers with 128 KiB Flash but only 16 KiB RAM. I wouldn't want to waste any RAM in such a scenario.

Is the situation where size_t is smaller than uintptr_t valid? I'm thinking of code which calculates a memory size by substracting 2 pointers. Would this code be able to assume then that the difference fits in a size_t? I tried to look it up, but I am getting confused by statements like this: "If an array is so large (greater than PTRDIFF_MAX elements, but less than SIZE_MAX bytes), that the difference between two pointers may not be representable as std::ptrdiff_t, the result of subtracting two such pointers is undefined."

Wouldn't it be cleaner to stop using structures which are prone to automatic padding and calculate the address of the size field in code? I don't really think the aim should be to minimise changes in the rest of the code.

**Comment by keith-packard** [comment]
> I still don't quite grasp it. Inserting padding would potentially waste memory, wouldn't it? Let's say the size and alignment for a pointer is 64 bits, and the size and alignment for size_t are 32 bits. Shouldn't the aim be to avoid the automatic padding? If the 64-bit pointer is aligned, the 32-bit size_t immediately before it would be sufficiently aligned too, right?

Yup, the problem is that the struct *will* have padding between `size` and `next` if `size` is smaller than `next`. A perfect fix would eliminate the struct entirely and use pointer arithmetic and a pile of casts to fetch the size placed just below the chunk. That's a much bigger change and only benefits systems with `size_t` smaller than a pointer. The only target I know of like that is the 8051 where pointers are three bytes while `size_t` is two bytes.

> I'm thinking of some weird 8-bit or 16-bit architecture with much more ROM/Flash than RAM. Pointers may be longer, in order to accommodate for the different memory areas, but dynamic allocations in RAM could be much smaller. I have STM32 microcontrollers with 128 KiB Flash but only 16 KiB RAM. I wouldn't want to waste any RAM in such a scenario.

Yeah, but ARM has 4 byte pointers and 4 byte `size_t`. Building a library with 2 byte sizes would certainly be possible, but that would require hacking the code.

> Is the situation where size_t is smaller than uintptr_t valid? I'm thinking of code which calculates a memory size by substracting 2 pointers.

Yup, the 8051 lives in this world -- sdcc uses three bytes for pointers -- one byte for the 'space' (text/small-data/large-data) and two for the value -- but memory is limited to 64kB, so size_t is two bytes (I think).

> Would this code be able to assume then that the difference fits in a size_t? I tried to look it up, but I am getting confused by statements like this: "If an array is so large (greater than PTRDIFF_MAX elements, but less than SIZE_MAX bytes), that the difference between two pointers may not be representable as std::ptrdiff_t, the result of subtracting two such pointers is undefined."

Welcome to C? `ptrdiff_t` is signed, while `size_t` is unsigned, and signed overflow is undefined.

> Wouldn't it be cleaner to stop using structures which are prone to automatic padding and calculate the address of the size field in code? I don't really think the aim should be to minimise changes in the rest of the code.

Agreed. However, we can fix the current implementation without requiring a perfect solution. If you want to try your hand at fixing this for real, I'd be super happy to help get that working.

**Comment by arichardson** [comment]
> Hrm. The proposed change (`size_t` -> `uintptr_t`) assumes that `uintptr_t` and `struct malloc_chunk *` have the same size. That's likely true in any "reasonable" C implementation, but I don't find any assurances about that in the standard.

The standard requires that pointers can round-trip via `uintptr_t`, so the only reasonable approach is for uintptr_t to be able to hold all possible bit values of a pointer, which implies that it is at least as big as `void*`. Of course uintptr_t is not strictly required for all versions of C, but if it exists those properties must hold.

**Comment by keith-packard** [comment]
> The standard requires that pointers can round-trip via `uintptr_t`, so the only reasonable approach is for uintptr_t to be able to hold all possible bit values of a pointer, which implies that it is at least as big as `void*`. Of course uintptr_t is not strictly required for all versions of C, but if it exists those properties must hold.

On 8051 using sdcc, uintptr_t is larger than a pointer (uintptr_t is four bytes, pointers are three bytes) while size_t is smaller (two bytes). Of course, the alignment required for each of them is only one byte. So, on that target using the union of size_t with pointer saves memory compared to using uintptr_t. It still wastes a byte as every type need only be aligned on a byte boundary, so things could be even better if we restructure the code to eliminate `struct malloc_chunk` as that's what's forcing the unnecessary alignment between the chunk size and the free list next pointer. But that's a bunch more work, so let's focus on making it conform to the C standard (which the current code fails to do) and *then* improve it further.

In any case, do you agree that using `union { size_t size; struct malloc_chunk *pointer }` as both the alignment requirement (via _Alignof) and within the `struct malloc_chunk` itself in place of `size` meets your requirements? That seems effectively the same as what you've done with `uintptr_t` except that it makes `struct malloc_chunk` uses the space of (at least) a pointer for the header rather than the space of `uintptr_t`.

**Comment by keith-packard** [comment]
See what you think of #691; that avoids the whole intra-struct padding issue by not having a struct containing the size and next pointer. It's not as easy to read though.

**Comment by arichardson** [comment]
> See what you think of #691; that avoids the whole intra-struct padding issue by not having a struct containing the size and next pointer. It's not as easy to read though.

Sorry for the delay here been busy with other tasks. I'll go test your pr shortly. In the mean time I could drop the last two changes from the pr and only keep the first cleanup commits?

**Comment by keith-packard** [comment]
> > See what you think of #691; that avoids the whole intra-struct padding issue by not having a struct containing the size and next pointer. It's not as easy to read though.
> 
> Sorry for the delay here been busy with other tasks. I'll go test your pr shortly. In the mean time I could drop the last two changes from the pr and only keep the first cleanup commits?

I've merged those changes into #691. I can pull over your patches and rebase my other changes on top of them if you'd like?

**Comment by arichardson** [comment]
> > > See what you think of #691; that avoids the whole intra-struct padding issue by not having a struct containing the size and next pointer. It's not as easy to read though.
> > 
> > 
> > Sorry for the delay here been busy with other tasks. I'll go test your pr shortly. In the mean time I could drop the last two changes from the pr and only keep the first cleanup commits?
> 
> I've merged those changes into #691. I can pull over your patches and rebase my other changes on top of them if you'd like?

That sounds ideal. I'll hopefully find time this afternoon/evening to test your PR on a CHERI system.

**Comment by keith-packard** [comment]
> That sounds ideal. I'll hopefully find time this afternoon/evening to test your PR on a CHERI system.

Done. Lmk what you think; it's not as easy to read as having the size and next pointer in the same struct, but it reflects how things work more accurately.

**Comment by keith-packard** [comment]
With the merge of #691, I'm going to close this PR.
